[jit] Add RRef to IValue and JIT type system

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33263 [rpc] allow RRef local creation with IValue objects
* #33190 [jit] Initial use RRef in TorchScript
* **#32992 [jit] Add RRef to IValue and JIT type system**
* #33189 [rpc] Switch RRef to be managed by intrusive_ptr

This PR add RRef to IValue and the JIT type system.

- The RRefInterface abstract class inherit from intrusive_ptr_target,
  this made the RRef class can be hold in ivalue as intrusive_ptr

- Add RRefType as a JIT type, it's a container type similar to
future type.

Differential Revision: [D19871242](https://our.internmc.facebook.com/intern/diff/D19871242)

wanchaol committed:
[jit] Add RRef to IValue and Jit

This PR add the Remote Reference type from the torch.distributed.rpc
package into IValue and JIT. We introduce an RRefInterface abstract
class and the base of RRef >: UserRRef/OwnerRRef, and made this inherit
from intrusive_ptr_target, and made the rref to be hold as IValue.

We also add RRefType as a JIT type, it's a container type similar to
future type.

Note that all of the rpc code in JIT is hidden behind the build flag

[ghstack-poisoned]

zhaojuanmao commented on torch/csrc/jit/register_distributed_ops.cpp:
```diff
@@ -23,6 +28,29 @@ at::Tensor optional_to_tensor(c10::optional<at::Tensor> v) {
+           if (rref->isOwner()) {
+              res = c10::dynamic_intrusive_pointer_cast<dist_rpc::OwnerRRef>(rref)->getValue();
+           } else {
+              res = c10::dynamic_intrusive_pointer_cast<dist_rpc::UserRRef>(rref)->toHere();
```
now toHere() is rawIValue, need to call res = std::move(rawValues).front() after this line

wanchaol replied:
I changed the UserRRef C++ API to still return a single IValue instead of a std::vector<IValue>. IValue should itself handle the list abstraction.

zhaojuanmao commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1711,3 +1711,23 @@ def test_function_not_on_callee(self):
+            # type: (RRef[Tensor]) -> Tensor
+            return rref_var.to_here()
+
+        res = rref_to_here(rref_var)
```
do you want to assertEqual to validate the result?

wanchaol committed:
Update on "[jit] Add RRef to IValue and Jit"


This PR add the Remote Reference type from the torch.distributed.rpc
package into IValue and JIT. We introduce an RRefInterface abstract
class and the base of RRef >: UserRRef/OwnerRRef, and made this inherit
from intrusive_ptr_target, and made the rref to be hold as IValue.

We also add RRefType as a JIT type, it's a container type similar to
future type.

Note that all of the rpc code in JIT is hidden behind the build flag

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Add RRef to IValue and Jit"


This PR add the Remote Reference type from the torch.distributed.rpc
package into IValue and JIT. We introduce an RRefInterface abstract
class and the base of RRef >: UserRRef/OwnerRRef, and made this inherit
from intrusive_ptr_target, and made the rref to be hold as IValue.

We also add RRefType as a JIT type, it's a container type similar to
future type.

Note that all of the rpc code in JIT is hidden behind the build flag

[ghstack-poisoned]

suo reviewed (changes requested):
An initial request: I started taking a look at this PR, but it seems like it would be faster to review if we split it up into a stack with each PR doing a single thing. For example: 
- Switching `RRefInterface` from `shared_ptr` to `intrusive_ptr`
- Adding the basic JIT type and ivalue for rref
- Adding the actual use of the JIT type in the system (including the serialization/pybind stuff

zzzwen commented:
Can we add Python3 type support
List[rpc.RRef[torch.Tensor]]
Will produce error
TypeError: 'pybind11_type' object is not subscriptable

wanchaol committed:
Update on "[jit] Add RRef to IValue and Jit"


This PR add the Remote Reference type from the torch.distributed.rpc
package into IValue and JIT. We introduce an RRefInterface abstract
class and the base of RRef >: UserRRef/OwnerRRef, and made this inherit
from intrusive_ptr_target, and made the rref to be hold as IValue.

We also add RRefType as a JIT type, it's a container type similar to
future type.

Note that all of the rpc code in JIT is hidden behind the build flag

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Add RRef to IValue and JIT type system"


This PR add RRef to IValue and the JIT type system.

- The RRefInterface abstract class inherit from intrusive_ptr_target,
  this made the RRef class can be hold in ivalue as intrusive_ptr

- Add RRefType as a JIT type, it's a container type similar to
future type.

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Add RRef to IValue and JIT type system"


This PR add RRef to IValue and the JIT type system.

- The RRefInterface abstract class inherit from intrusive_ptr_target,
  this made the RRef class can be hold in ivalue as intrusive_ptr

- Add RRefType as a JIT type, it's a container type similar to
future type.

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Add RRef to IValue and JIT type system"


This PR add RRef to IValue and the JIT type system.

- The RRefInterface abstract class inherit from intrusive_ptr_target,
  this made the RRef class can be hold in ivalue as intrusive_ptr

- Add RRefType as a JIT type, it's a container type similar to
future type.

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Add RRef to IValue and JIT type system"


This PR add RRef to IValue and the JIT type system.

- The RRefInterface abstract class inherit from intrusive_ptr_target,
  this made the RRef class can be hold in ivalue as intrusive_ptr

- Add RRefType as a JIT type, it's a container type similar to
future type.

Differential Revision: [D19871242](https://our.internmc.facebook.com/intern/diff/D19871242)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Add RRef to IValue and JIT type system"


This PR add RRef to IValue and the JIT type system.

- The RRefInterface abstract class inherit from intrusive_ptr_target,
  this made the RRef class can be hold in ivalue as intrusive_ptr

- Add RRefType as a JIT type, it's a container type similar to
future type.

Differential Revision: [D19871242](https://our.internmc.facebook.com/intern/diff/D19871242)

[ghstack-poisoned]
